api/python/provisioner/commands/auto_deploy.py

```diff
@@ -61,12 +61,15 @@ def deployment_validations(deploy_check):
             raise ValueError("Error During Deployment "
                              f"{deploy_check} Validations: {str(exc)}")
         else:
-            if "pre" in deploy_check:
+            if deploy_check == GroupChecks.DEPLOY_PRE_CHECKS.value:
                 check_import.PreChecksDecisionMaker().make_decision(
                     check_result=check_res)
-            else:
+            elif deploy_check == GroupChecks.DEPLOY_POST_CHECKS.value:
                 check_import.PostChecksDecisionMaker().make_decision(
                     check_result=check_res)
+            else:
+                raise ValueError(
+                    f'Group Check "{deploy_check}" is not supported')
```

Comment:
Overall, looks good to me. 
Although, we're calling the pre-check or post-check before and after Deploy step explicitly with the same GroupCheck value. 
`self.deployment_validations(GroupChecks.DEPLOY_PRE_CHECKS.value)`

So, validating them again here and raising an `else` block for exception may be unnecessary. Just my thought @mazinamdar @dmitrydb

Reply:
Yes, at the first glance, it seems like unnecessary. But I always can do this:
```python
from .auto_deploy import AutoDeploy

AutoDeploy.deployment_validations("Hello")
```
And without this changes it will execute `PostChecksDecisionMaker` validation.
So as result, Mazhar's changes helps to keep method according to designed flow.

Reply:
Sure, but any sample name in the check_name will be handled in Exception in `check.run()` .
But, okay, let us have the `else` block. Thanks

```
        try:
            check_res = check_cmd().run(deploy_check)
        except Exception as exc:
            raise ValueError("Error During Deployment "
                            f"{deploy_check} Validations: {str(exc)}")
```

Reply:
@divya-vijay I think we don't understand each other. Let me just clarify what I did mean.
My previous example is not the best choice to demonstrate my point. 
So before this `else` block appeared, I could call
```python
from .auto_deploy import AutoDeploy

AutoDeploy.deployment_validations("swupdate_checks")
```
As it is the correct and supported value from `GroupChecks`, there will be no Exceptions here:
```
        try:
            check_res = check_cmd().run(deploy_check)
        except Exception as exc:
            raise ValueError("Error During Deployment "
                            f"{deploy_check} Validations: {str(exc)}")
```

But after that we had the following block (before new and latest Mazhar changes)
```python
            if "pre" in deploy_check:
                check_import.PreChecksDecisionMaker().make_decision(check_result=check_res)
            else:
                check_import.PostChecksDecisionMaker().make_decision(check_result=check_res)
```
in case of `swupdate_checks` parameter `else` block will be executed.
So question to you @divya-vijay. Do you think that this is the correct behavior? If you think that was the plan, I don't mind reverting changes.

Reply:
Based on this conversation, I would recommend to implement it as:
```python

_DECISION_MAKERS = {
    "deploy_pre_checks": "PreChecksDecisionMaker",
    "deploy_post_checks": "PostChecksDecisionMaker",
}

@staticmethod
def deployment_validations(deploy_check):

    if deploy_check not in self._DECISON_MAKERS:
        raise ValueError(f'Group Check "{deploy_check}" is not supported')

    check_import = importlib.reload(check)

    check_cmd = check_import.Check

    try:
        check_res = check_cmd().run(deploy_check)
    except Exception as exc:
        raise ValueError("Error During Deployment "
                         f"{deploy_check} Validations: {str(exc)}")


    getattr(check_import, deploy_check)().make_decision(check_result=check_res)
```

Reply:
@dmitrydb No, I understood what you meant. 
What I meant was, we have a similar implementation in `replace_node.py` .
`def repl_node_validations(repl_check):`
We initially decided to make this function common so all the modules can use from a single place, but later decided to have specific it as per each implementation. In a common implementation, validating the incoming param is required. 
But now, the `check` parameter that is passed to this method would already be specific to it and validated. 
Which is why I thought re-validating them with the same `value == GroupChecks.DEPLOY_PRE_CHECKS.value` might not be necessary. 
As long as the check name is valid, we only need to verify if they belong to either pre- or post- category. 

In case of your example, "swupdate_checks" does not have "pre" in it, so, post-checks block will be executed and raise warning (which is how we're handling checks for SW now). 

But I understood your implementation suggestion here. So, we can have the same for now as per Mazhar's PR, and based on more modules validations in future, we can add any more improvements to it.

Reply:
>What I meant was, we have a similar implementation in `replace_node.py`.
`def repl_node_validations(repl_check):`

I have the same concern regarding to `repl_node_validations(repl_check)`. Why do we need to pass `repl_check` parameter to this function? I think this method can be presented as:
```python
    @staticmethod
    def repl_node_validations():
        checker = Check()
        try:
            check_res = checker.run(GroupChecks.REPLACENODE_CHECKS.value)
        except Exception as exc:
            raise errors.ValidationError("Error During Replace Nodes "
                                                       f"replacenode_checks Validations: {str(exc)}")
        else:
            PreChecksDecisionMaker().make_decision(check_result=check_res)
```
In this implementation, you don't need to care about parameter validations and anything else because this method does just one thing: runs the desired group of checks for replace_node verification and executes the `PreChecks` decision-maker.

>We initially decided to make this function common so all the modules can use from a single place, but later decided to have it specific as per each implementation. In a common implementation, validating the incoming param is required.
But now, the check parameter that is passed to this method would already be specific to it and validated.
Which is why I thought re-validating them with the same value == GroupChecks.DEPLOY_PRE_CHECKS.value might not be necessary.

I can't agree with this assumption that for common implementation we need to validate incoming parameters, but for specific implementations, we don't. If you have no parameters validation for a specific implementation, so it means that you have just a common implementation. Because it can be executed with all parameters that `check_res = check_cmd().run(deploy_check)` accepts including single checks (**"network", "hostnames", "mgmt_vip"**, etc.). And for all those cases we will run the `PostChecksDecisionMaker` decision-maker class.  So my main concern here if you mark class method to be specific, so it should have specific behavior and should not allow executing cases which I mentioned even if we are thinking that this method will be invoked always correctly and with correct input parameters.

>In case of your example, "swupdate_checks" does not have "pre" in it, so, post-checks block will be executed and raise warning (which is how we're handling checks for SW now).

Yes, exactly. For `swupdate_checks` we will run `PostChecksDecisionMaker` class which works the same as `SWUpdateDecisionMaker`.
But again, consider this case
```python
from .auto_deploy import AutoDeploy

AutoDeploy.deployment_validations("replacenode_checks")
```
In `def repl_node_validations(repl_check):` we execute in this case `PreChecksDecisionMaker().make_decision(check_result=check_res)`. 
But for `def deployment_validations(deploy_check):` we execute `check_import.PostChecksDecisionMaker().make_decision(check_result=check_res)`. 

`PreChecksDecisionMaker` raises exceptions for fail results, `PostChecksDecisionMaker` shows warning message to user. So for me it looks like a conflict here. So this is why I requested Mazhar to add `else` block here.

To conclude our long and productive discussion, we have three options here:
* We can revert changes for `AutoDeploy.deployment_validations`
* We can keep the current variant
* We can change to use this variant: https://github.com/Seagate/cortx-prvsnr/pull/778#discussion_r534212931

So, @mazinamdar, it is up to you which option to choose. 

I think this variant: https://github.com/Seagate/cortx-prvsnr/pull/778#discussion_r534212931
is more readable and simple for the user's understanding.

Reply:
Thank you, Dmitry! Yes, I agree with the conclusion options here. Based on time as a factor, we can choose either of second or third option.